taar_api/api/views.py

```diff
@@ -1,15 +1,22 @@
+from django.core.cache import cache
 from django.conf import settings
 from django.http import JsonResponse
 
-from taar.recommenders import RecommendationManager
+from taar import recommenders
 from taar.profile_fetcher import ProfileFetcher
 from taar.hbase_client import HBaseClient
 
 
 def recommendations(request, client_id):
     """Return a list of recommendations provided a telemetry client_id."""
-    hbase_client = HBaseClient(settings.HBASE_HOST)
-    profile_fetcher = ProfileFetcher(hbase_client)
-    recommendation_manager = RecommendationManager(profile_fetcher)
+    recommendation_manager = cache.get("recommendation_manager")
+    if recommendation_manager is None:
+        hbase_client = HBaseClient(settings.HBASE_HOST)
+        profile_fetcher = ProfileFetcher(hbase_client)
+        recommendation_manager = recommenders.RecommendationManager(profile_fetcher)
+        # Cache the recommendation manager for 1h
+        cache.set("recommendation_manager",
+                  recommendation_manager,
+                  3600)
```

Comment:
That's an interesting approach. So, please correct me if I'm wrong.

*Current state*

- Every time a request is made, we instantiate a new `RecommendationManager`. However, we don't get too much of a performance hit because we cache the data files locally, so we don't hit network but rather disk I/O.
- `taar-lib `cache files don't get invalidated unless the the machine restarts.

*With this PR*

- The `RecommendationManager`, along with any model, gets cached by Django for 1hr.
- All the requests, within the hour, will be served by the same manager.
- After 1hr, the cache gets invalidated. The next request will instantiate a new manager.
- However the cached data files for the models will still be on the disk, so they won't be refreshed.

I think we can safely cache for a few hours and delegate all the caching responsibilities to `taar-api`: we could change `taar-lib` to be dumb and not cache. What do you think?

Reply:
> Every time a request is made, we instantiate a new RecommendationManager. However, we don't get too much of a performance hit because we cache the data files locally, so we don't hit network but rather disk I/O.

This is not entirely correct. [We don't always cache results locally](https://github.com/mozilla/taar/blob/0802bfad8aae984101b4b088850cebb25866e9ad/taar/recommenders/collaborative_recommender.py#L38-L42). And even if we cached that on disk, we would still have to come up with an expiration mechanism.

> taar-lib cache files don't get invalidated unless the the machine restarts.

If we taar-lib cache you mean the content downloaded using `get_s3_json_content` that's correct, if the docker container gets destroyed the file-based cache is gone.

> However the cached data files for the models will still be on the disk, so they won't be refreshed.

That's true for the content fetched by `get_s3_json_content`.

Ideally we could pass a cache object down to the recommenders and, in the context of taar_api, that object would be backed by a django cache obj.